### ISSUE TITLE
Local-search phase for CVRP

### ISSUE THREAD

**Issue by jcoupey** [VROOM-Project/vroom#92]
The current CVRP heuristic implementation is detailed in #74 and implemented in #75. It's really just a heuristic approach that logically suffers from biases.

Looking at solution samples make it obvious that we could come up with a few operators to try and improve the heuristic solutions during a local-search phase.

I want to keep the searching time low. I think a good start would be (as for the TSP local search) to implement a few operations and just design a greedy choice scheme for fast convergence.

**Comment by jcoupey** [comment]
Update on what's going on in #110.

# Operators

I implemented a bunch of moves that operate on pairs of routes. Four first will only swap nodes/edges while the last two can swap whole route parts. They all inherit from `ls_operator` that provide virtual `is_valid`, `gain` and `apply` member functions.

Pictures below describe the idea behind each operator, actual implementation is a bit more complex due to edge cases handling (e.g. open tours). Cost evaluation gets a speedup from pre-computing (and updating) several costs/amount properties that are often used.

## Relocate

![relocate](https://user-images.githubusercontent.com/6639205/41353400-721df1f4-6f1c-11e8-9e1a-cfd747fbfdf2.png)

## Exchange

![exchange](https://user-images.githubusercontent.com/6639205/41353411-77ddfbd4-6f1c-11e8-89bc-876b52484d73.png)

## Or-opt

![or_opt](https://user-images.githubusercontent.com/6639205/41353420-7ddc1ce6-6f1c-11e8-870a-b48bcc76e432.png)

Swapped edge can be reversed or not.

## CROSS-Exchange

![cross_exchange](https://user-images.githubusercontent.com/6639205/41353427-84325650-6f1c-11e8-89be-d236cc1a19bd.png)

Swapped edges can be reversed or not.

## 2-opt*

![2-opt](https://user-images.githubusercontent.com/6639205/41353438-8852c512-6f1c-11e8-972d-3554fe57c1a3.png)

## Reverse 2-opt*

![reversed_2-opt](https://user-images.githubusercontent.com/6639205/41353445-8dfb01be-6f1c-11e8-8f37-282641c07186.png)

# Local search scheme

## Local search step

We basically try all of the above, storing the best improvement option per route pair. Then we apply the best improvement overall and iterate until there is no possible improvement left.

The good thing is that operator application only invalidates previously known operations that are related to the two used routes. Only those are recomputed, so most of the best known improvements are kept after the first step. The number of pairs of routes to deal with is quadratic in the number of vehicles for the first step, then linear for all subsequent steps.

## Getting out of a "bad" local minimum

Because the strategy is a gradient descent, we'll just end up in a local minimum. Sometimes there are "obviously" wrong things in the solution that can't be undone with the above moves, usually due to the tightness of the capacity constraints. We try to get out of that local minimum by removing the best candidate in each route, then including them back in a (hopefully) more clever way. After that we head for a new local search step, and repeat until this process does not improve the best known solution.

- Best candidate is decided based on the removal gain and the reaching cost from any of the other compatible current routes.
- Job inclusion is done based on addition cost pondered by some regret (cost of having to add the job later in another route).

This yields better solutions and sometimes also make it possible to include more jobs. On the other end, it roughly doubles the local search computing time, so it might be a good idea to provide a high-level knob to have that as an optional feature.

# Parallelization and computing times

We run the local search in parallel on each of the initial solutions provided by the clustering heuristic. Overall computing time is directly a factor of the number of different parameters used for clustering (probably still too much right now). Tuning this is out of scope here and should be handled in #120.

**Comment by jcoupey** [comment]
Sneak peek on performance, based on current branch `feature/CVRP-local-search` at 
40635ac. I've used the 189 instances from the A, B, E, F, M, P, and X classes in CVRPLIB, with sizes ranging from 15 to 1000 jobs.

- Total computing time is around 45 minutes (can probably easily be brought way down through #120)
- 44922 out of the 44993 jobs are assigned (that's 99.8%)
- 166 out of the 189 instances are solved completely (no unassigned job)
- On those 166 instances where it makes sense to compare the cost to the best known value (usually optimal):
  - more than 10% of instances (19) are solved to the best known cost
  - the average gap is +3.0%
  - the median gap is +2.4%
  - the worst-case gap is +15.1%

**Comment by jcoupey** [comment]
Landed with #110.
